core/src/main/scala/kafka/utils/ZkUtils.scala

```diff
@@ -256,19 +256,39 @@ class ZkUtils(val zkClient: ZkClient,
   }
 
   /**
-   * Register brokers with v2 json format (which includes multiple endpoints).
+   * Register brokers with v3 json format (which includes multiple endpoints and rack).
    * This format also includes default endpoints for compatibility with older clients.
+   *
    * @param id
    * @param host
    * @param port
    * @param advertisedEndpoints
    * @param jmxPort
    */
-  def registerBrokerInZk(id: Int, host: String, port: Int, advertisedEndpoints: immutable.Map[SecurityProtocol, EndPoint], jmxPort: Int) {
+  def registerBrokerInZk(id: Int,
+                         host: String,
+                         port: Int,
+                         advertisedEndpoints: immutable.Map[SecurityProtocol, EndPoint],
+                         jmxPort: Int,
+                         rack: Option[String] = Option.empty,
+                         apiVersion: ApiVersion = ApiVersion.latestVersion) {
     val brokerIdPath = BrokerIdsPath + "/" + id
     val timestamp = SystemTime.milliseconds.toString
 
-    val brokerInfo = Json.encode(Map("version" -> 2, "host" -> host, "port" -> port, "endpoints"->advertisedEndpoints.values.map(_.connectionString).toArray, "jmx_port" -> jmxPort, "timestamp" -> timestamp))
+    val version = if (apiVersion >= KAFKA_0_10_0_IV0) 3 else 2
+    var jsonMap = Map("version" -> version,
+                      "host" -> host,
+                      "port" -> port,
+                      "endpoints"->advertisedEndpoints.values.map(_.connectionString).toArray,
+                      "jmx_port" -> jmxPort,
+                      "timestamp" -> timestamp
+    )
+    rack match {
+      case Some(rack) if (version >= 3) => jsonMap += ("rack" -> rack)
```

Comment:
I don't really know what would be the valid characters or length limit for rack. Looking at the implementation of Json.encode() there is nothing suspicious how characters are handled.

Note that in some cases, rack can be a logical name and used for grouping brokers for fault tolerance. So any character is possible.

Apache Cassandra does not seem to do any validation on rack name for their property file based configuration.

If there is no specification or usual convention for the rack name, I suggest we leave it unchecked.